# Project 1: Shark attacks

## Cleaning and wrangling

### 1. Open the document

In [1]:
import pandas as pd

In [2]:
attacks = pd.read_csv("../data/attacks.csv", encoding = "latin-1")

In [3]:
attacks.head()

Case Number         Date    Year        Type    Country             Area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   ...  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F  ...   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F  ...   
2                    Habush, Oahu      Surfing      John Denges    M  ...   
3              Arrawarra Headland      Surfing             male    M  ...   
4                        La Ticla  Free diving   Gustavo Ramos     M  ...   

          Species           Investigator or Source                       pdf  \
0      White shark                R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1              NaN  K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2              NaN  K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3        2 m shark                  B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4  Tiger shark, 3m                       A .Kipper      2018.06.04-Ramos.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.04   

  Case Number.2 original order Unnamed: 22 Unnamed: 23  
0    2018.06.25         6303.0         NaN         NaN  
1    2018.06.18         6302.0         NaN         NaN  
2    2018.06.09         6301.0         NaN         NaN  
3    2018.06.08         6300.0         NaN         NaN  
4    2018.06.04         6299.0         NaN         NaN  

[5 rows x 24 columns]

### 2. Exploring data

#### 2.1. Size of the dataset

Let's take a look at the size of the dataset:

In [4]:
attacks.shape

(25723, 24)

There are 25723 records of shark attacks in our dataset. In each record, 24 characteristics were saved. Which are these 24 characteristics?

In [5]:
attacks.columns

Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species ', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

#### 2.2. Duplicates

There could be some duplicates which should be eliminated from the dataset.

In [6]:
attacks = attacks.drop_duplicates()
attacks.shape

(6312, 24)

There were 19411 duplicates, which included empty rows.

#### 2.2. Empty rows and columns

**Is there any record (row) empty?**

In [7]:
emptyrows = attacks.isnull().sum(axis=1) #count number of empty data in each row
print(sum(emptyrows ==24)) #counts number of empty rows
print(sum(emptyrows ==23)) #counts number of rows with only one data
print(sum(emptyrows ==22)) #counts number of rows with only two data

1
2
7


We have eliminated all empty rows except the first one. There are also records with only one or two data which doesn't help us to solve the hypothesis. Let's check wich data are completed in those records which only have two data:

In [8]:
attacks[emptyrows==22]

Case Number Date  Year Type Country Area Location Activity Name Sex   \
6302           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
6303           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
6304           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
6305           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
6306           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
6307           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
6308           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   

      ... Species  Investigator or Source  pdf href formula href  \
6302  ...      NaN                    NaN  NaN          NaN  NaN   
6303  ...      NaN                    NaN  NaN          NaN  NaN   
6304  ...      NaN                    NaN  NaN          NaN  NaN   
6305  ...      NaN                    NaN  NaN          NaN  NaN   
6306  ...      NaN                    NaN  NaN          NaN  NaN   
6307  ...      NaN                    NaN  NaN          NaN  NaN   
6308  ...      NaN                    NaN  NaN          NaN  NaN   

     Case Number.1 Case Number.2 original order Unnamed: 22 Unnamed: 23  
6302           NaN           NaN         6304.0         NaN         NaN  
6303           NaN           NaN         6305.0         NaN         NaN  
6304           NaN           NaN         6306.0         NaN         NaN  
6305           NaN           NaN         6307.0         NaN         NaN  
6306           NaN           NaN         6308.0         NaN         NaN  
6307           NaN           NaN         6309.0         NaN         NaN  
6308           NaN           NaN         6310.0         NaN         NaN  

[7 rows x 24 columns]

No useful information can be extracted from those rows either so, let's delete rows that have 2 or less data.

In [9]:
isemptyrow = [i for i in attacks.index if emptyrows[i] >= 22] #list with the indexes of the empty rows
attacks = attacks.drop(isemptyrow) #df with at least one data for each record

In [10]:
attacks.shape

(6302, 24)

We have reduced the dataset to 6302 records.

**How many nulls do we have in each column?**

In [11]:
attacks.isnull().sum()

Case Number                  1
Date                         0
Year                         2
Type                         4
Country                     50
Area                       455
Location                   540
Activity                   544
Name                       210
Sex                        565
Age                       2831
Injury                      28
Fatal (Y/N)                539
Time                      3354
Species                   2838
Investigator or Source      17
pdf                          0
href formula                 1
href                         0
Case Number.1                0
Case Number.2                0
original order               0
Unnamed: 22               6301
Unnamed: 23               6300
dtype: int64

We don't have any data in columns "Unnamed: 22" and "Unnamed: 23" as the number of nulls is equal to the number of rows so let's delete them too.

In [12]:
attacks = attacks.drop(columns = ["Unnamed: 22", "Unnamed: 23"])

From the other hand, columns named "Case Number", "Case Number.1" and "Case Number.2" and "Date" seem to have the same values, although in different format:

In [13]:
print(attacks["Case Number"][0])
print(attacks["Case Number.1"][0])
print(attacks["Case Number.2"][0])
print(attacks["Date"][0])

2018.06.25
2018.06.25
2018.06.25
25-Jun-2018


Let's change the format of attacks["Date"].

In [14]:
def dateformat (date):  #Creo que se puede hacer con regex
    [dd, mm, yyyy] = date.split('-')
    months1=["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sept", "Oct", "Nov", "Dec"]
    months2=["01", "02", "03", "04", "05", "06", "07", "08", "09", "10", "11", "12"]
    for m1,m2 in zip(months1, months2):
        if mm == m1: mm=m2
    return '.'.join([yyyy,mm,dd])

In [15]:
for i in attacks.index:
    original = attacks["Date"][i]  
    try:
        attacks["Date"][i] = dateformat(original)
    except:
        pass

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


And now, values from columns "Case Number", "Case Number.1", "Case Number.2" and "Date" are compared.

In [24]:
CN_CN1 = (attacks["Case Number"] == attacks["Case Number.1"])
CN_CN2 = (attacks["Case Number"] == attacks["Case Number.2"])
CN1_CN2 = (attacks["Case Number.1"] == attacks["Case Number.2"])
CN2_D = (attacks["Case Number.2"] == attacks["Date"])
print('Equal values in Case Number and Case N1: ', sum(CN_CN1))
print('Equal values in Case Number and Case N2: ', sum(CN_CN2))
print('Equal values in Case N1 and Case N2: ', sum(CN1_CN2))
print('Equal values in Case N2 and Date: ', sum(CN2_D))

Equal values in Case Number and Case N1:  6278
Equal values in Case Number and Case N2:  6298
Equal values in Case N1 and Case N2:  6282
Equal values in Case N2 and Date:  3258


In [25]:
rowsCN = CN_CN1 + CN_CN2 + CN1_CN2 == False

/home/b/.local/lib/python3.6/site-packages/pandas/core/computation/expressions.py:178: UserWarning: evaluating in Python space because the '+' operator is not supported by numexpr for the bool dtype, use '|' instead
  f"evaluating in Python space because the {repr(op_str)} "


Let's fix first the differences between "Case Number", "Case Number.1", "Case Number.2".

In [26]:
attacks[["Case Number","Case Number.1", "Case Number.2"]][rowsCN]

Empty DataFrame
Columns: [Case Number, Case Number.1, Case Number.2]
Index: []

* Sustituir valores nulos